### PR TITLE
deprecate compat & encoding

### PR DESCRIPTION
**Pull request by max-sixty**
Still need to adjust the tests

 - [x] Closes https://github.com/pydata/xarray/issues/1188
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -194,13 +192,14 @@ def __init__(self, data, coords=None, dims=None, name=None,
...
-            include '_FillValue', 'scale_factor', 'add_offset', 'dtype',
-            'units' and 'calendar' (the later two only for datetime arrays).
-            Unrecognized keys are ignored.
+        encoding : deprecated
         """
+
+        if encoding is not None:
+            warnings.warn(
+                'The `encoding` argument to `DataArray` is deprecated. '
+                'Instead, specify the encoding when writing to disk.',
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Let's also mention "or set the `encoding` attribute directly."

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -194,13 +192,14 @@ def __init__(self, data, coords=None, dims=None, name=None,
...
+                FutureWarning)
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Be sure to specify `stacklevel` here (e.g., `stacklevel=2`) so users get a warning at the appropriate place in their code.

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -346,16 +346,16 @@ def __init__(self, data_vars=None, coords=None, attrs=None,
...
-            - 'identical': all values, dimensions and attributes must be the
-              same.
+        compat : deprecated
         """
+
+        if compat is not None:
+            warnings.warn(
+                'The `compat` argument to Dataset is deprecated. '
+                'Instead, use `merge` to control how variables are combined',
+                FutureWarning)
```
Comment (shoyer) [inline xarray/core/dataset.py]: Likewise, set `stacklevel` here.

**Comment by max-sixty** [comment]
Green!

**Comment by max-sixty** [comment]
> Green!

Please could someone review and approve?

File: xarray/core/dataarray.py [hunk 4]
```diff
@@ -192,13 +190,15 @@ def __init__(self, data, coords=None, dims=None, name=None,
...
-            Unrecognized keys are ignored.
+        encoding : deprecated
         """
+
+        if encoding is not None:
+            warnings.warn(
+                'The `encoding` argument to `DataArray` is deprecated. '
+                'Instead, specify the encoding when writing to disk or '
+                'set the `encoding` attribute directly.',
+                FutureWarning, stacklevel=2)
```
Comment (jhamman) [inline xarray/core/dataarray.py]: Should we mention that this will be removed in 0.13 or something?

Comment (max-sixty) [reply xarray/core/dataarray.py]: Yes, done

**Pull request merged**